### ISSUE TITLE
1.11.1: breaking change in extension loading

### ISSUE THREAD

**Issue by aplavin** [JuliaLang/julia#56204]
Not too often I see packages broken by patch versions of Julia, but it's one of those cases :(
As a specific example, `using UnionCollections, Dictionaries, FlexiMaps` throws `ERROR: LoadError: type Nothing has no field UnionDictionary` on this line: https://github.com/JuliaAPlavin/UnionCollections.jl/blob/0000000018ecd297a2e7448671da83c035b3da4c/ext/FlexiMapsDictionariesExt.jl#L8.

Effectively, it does `get_extension(MyPackage, :WeakDepA)` in the extension for WeakDepA + WeakDepB.
Judging by the documentation, it should work:
-  `get_extension` returns the extension if it's loaded
> get_extension(parent::Module, extension::Symbol)
> Return the module for extension of parent or return nothing if the extension is not loaded.

- Pkg docs say the following about extension loading:
> A package extension is a module in a file (similar to a package) that is automatically loaded when some other set of packages are loaded into the Julia session.

In the specific case with UnionCollections, this is what I expect to happen on that `using` line:
1. UnionCollections is loaded
2. Dictionaries is loaded
3. UnionCollections -> DictionariesExt is loaded (following Pkg docs)
4. FlexiMaps is loaded
5. UnionCollections -> FlexiMapsDictionariesExt is loaded, it sees DictionariesExt loaded already and successfully imports that (following get_extension docs)

And that's indeed how the observable behavior worked on earlier Julia versions, including 1.11.0. But not on 1.11.1.

**Comment by bvdmitri** [comment]
> Judging by the documentation, it should work:

AFAIK `get_extension` is internal (its also written in its docstring) and is subject to change and [its use is really discouraged](https://discourse.julialang.org/t/are-extension-packages-importable/92527/19) since the order load is not really defined.

**Comment by KristofferC** [comment]
Yeah, this is a bit unfortunate. The issue was that loading extensions in other extensions as was done previously was "unsound" and could easily cause "circular dependencies" leading to bad behavior like:

- https://github.com/JuliaLang/julia/issues/55557
- https://github.com/JuliaLang/julia/issues/55517
- https://github.com/JuliaLang/julia/issues/53081
- https://github.com/JuliaLang/julia/issues/52132

I didn't see any new regressions from PkgEval from backporting it which is why I assumed it wouldn't have a big impact on the ecosystem but maybe this was hidden by other errors. In hindsight it was probably wrong to backport it.

The use case people want seem to basically be https://github.com/JuliaLang/julia/issues/48734 so we should probably work on fixing that feature. In the mean time the question is if we revert the extension loading change for 1.11.2.

**Comment by KristofferC** [comment]
I looked in the PkgEval log and UnionCollections is here: https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/0de7b6c_vs_501a4f2/UnionCollections.primary.log. And we can see the extension fails to load but the package tests end up passing anyway...

**Comment by KristofferC** [comment]
I guess there is no test that checks that the function in the extension works?

**Comment by topolarity** [comment]
You might try using `eval` instead of `__precompile__(false)`:
```julia
# workaround for https://github.com/JuliaLang/julia/issues/56204
function __init__()
    eval(quote
        const DictionariesExt = Base.get_extension(UnionCollections, :DictionariesExt)
        FlexiMaps.mapview(f, d::DictionariesExt.UnionDictionary) = @modify(vs -> mapview(f, vs), DictionariesExt._values(d))

        # TODO disambiguation:
        # FlexiMaps.mapview(p::Union{Symbol,Int,String}, A::UnionDictionary) = mapview(PropertyLens(p), A)
    end)
end
```

**Comment by topolarity** [comment]
Never mind, that ends up eval'ing into a closed module so that's not allowed either.

I don't think either of these workarounds are viable right now...

`__precompile__(false)` is bugged for extensions, and `eval(...)` is not allowed to be deferred like I was trying to do

**Comment by topolarity** [comment]
I think the above workaround works with an appropriate pre-compile guard (the `!Base.generating_output()`)

**Comment by andreyz4k** [comment]
Another affected package is Transformers.js which relies on Flux's CUDA extension in its own CUDA extension. https://github.com/chengchingwen/Transformers.jl/blob/master/ext/TransformersCUDAExt/TransformersCUDAExt.jl

It compiles just fine but crashes in runtime with `ERROR: type Nothing has no field check_use_cuda`

**Comment by KristofferC** [comment]
A workaround would be to use Requires.jl style of conditional loading.

**Comment by kbarros** [comment]
This also seems to break precompilation of the [Sunny.jl](https://github.com/SunnySuite/Sunny.jl) package.

Sunny defines its plotting extensions in a module [PlottingExt](https://github.com/SunnySuite/Sunny.jl/blob/397b41c2ecbf1463762f86911ea7f9afc7226de3/ext/PlottingExt.jl) that depends on any variant of Makie (e.g. GLMakie or CarioMakie both pull in the Makie package, and therefore trigger loading of PlottingExt). Additionally, Sunny defines an extension module [GLMakiePrecompilesExt](https://github.com/SunnySuite/Sunny.jl/blob/397b41c2ecbf1463762f86911ea7f9afc7226de3/ext/GLMakiePrecompilesExt.jl) that precompiles code specific to the GLMakie backend, if it is loaded. Somehow this workflow became broken in 1.11.1, perhaps due to the order in which modules load.

Question: Is there some way to get precompilation working again on 1.11.1? Inspired by [a suggestion above](https://github.com/JuliaLang/julia/issues/56204#issuecomment-2419553167), I tried putting the precompile statements inside the guard:

```
if !Base.generating_output()
     Base.eval(quote
           PrecompileTools.@setup_workload ....
```

but in doing so, it seems that the precompile statements become ineffective (don't actually generate and cache the desired code).

**Comment by topolarity** [comment]
If #56368 lands, there will be a proper solution for this in 1.11.2:
> * Extensions may now depend on other extensions, if their triggers include all triggers of any
  extension they wish to depend upon (+ at least one other trigger). Ext-to-ext dependencies
  that don't meet this requirement are now blocked from using `Base.get_extension` during pre-
  compilation, to prevent extension cycles.

For example:
```julia
[extensions]
PlottingExt = "Plots"
StatisticsPlottingExt = ["Plots", "Statistics"]
```
With this, `StatisticsPlottingExt` is allowed to depend on `PlottingExt` (it can use get_extension during pre-compilation)

You can even do this across packages. For example, for the Transformers.jl + Flux.jl case above you'd need:
```julia
[extensions]
TransformersCUDAExt = ["Flux", "CUDA"]
```

TransformersCUDAExt triggers on `{Flux, CUDA, Transformers}` and FluxCUDAExt triggers on `{Flux, CUDA}`, so `get_extension` will work

**Comment by KristofferC** [comment]
> Sounds somewhat similar to "it will still have to come for Julia 2.0 anyway, so we are putting it into minor version already" 

Not really since the 1.10 behavior is buggy and would leave extensions in a state where their use should be actively discouraged since you at any point (from e.g. someone adding a dependency downstream) could cause your package to break.

**Comment by kbarros** [comment]
I tested on the [backports branch](https://github.com/JuliaLang/julia/pull/56228) for Julia 1.11.2 using `juliaup add pr56228`. It appears that this issue will be resolved as described in https://github.com/JuliaLang/julia/issues/56204#issuecomment-2442652997. Thank you everyone!

**Comment by topolarity** [comment]
This should now be resolved in 1.11.2, using the pattern / feature described in https://github.com/JuliaLang/julia/issues/56204#issuecomment-2442652997

**Comment by kbarros** [comment]
Initially, #56228 seemed to fix the [issue with precompiling package extensions](https://github.com/JuliaLang/julia/issues/56204#issuecomment-2439588043). Unfortunately, the fix appeared to stop working in the 1.11.2 release.

Maybe someone could help explain the proper workflow?

Concretely, our package Sunny now has in its Project.toml:

```
[extensions]
GLMakiePrecompilesExt = ["Makie", "GLMakie"]
PlottingExt = "Makie"
```

My understanding is that this would allow GLMakiePrecompilesExt to depend on methods in PlottingExt. GLMakiePrecompilesExt contains the following code:

```jl
import Sunny, GLMakie
import PrecompileTools as PT

PT.@setup_workload begin
    PT.@compile_workload begin
        # This specific method is defined in PlottingExt
        Sunny.view_crystal(...)
    end
end
```

Somewhere between the merge of #56228 and the release of 1.11.2, this workflow seemed to break. The error is:

```julia> using Sunny, GLMakie
Precompiling PlottingExt...
  1 dependency successfully precompiled in 8 seconds. 322 already precompiled.
Precompiling GLMakiePrecompilesExt...
Info Given GLMakiePrecompilesExt was explicitly requested, output will be shown live 
ERROR: LoadError: MethodError: no method matching view_crystal(::Sunny.Crystal)
The function `view_crystal` exists, but no method is defined for this combination of argument types.
```

This means that somehow GLMakiePrecompilesExt is not actually seeing the methods defined in PlottingExt. Any ideas?
